java-checks/src/main/java/org/sonar/java/checks/LazyArgEvaluationCheck.java

```diff
@@ -0,0 +1,294 @@
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.check.Rule;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.MethodMatcherCollection;
+import org.sonar.java.matcher.TypeCriteria;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.CatchTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Rule(key = "S2629")
+public class LazyArgEvaluationCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private static final String SLF4J_LOGGER = "org.slf4j.Logger";
+  public static final String JUL_LOGGER = "java.util.logging.Logger";
+  private static final TypeCriteria STRING = TypeCriteria.is("java.lang.String");
+  public static final TypeCriteria OBJECT_ARR = TypeCriteria.is("java.lang.Object[]");
+
+  interface LogLevels {
+    List<MethodMatcher> log();
+
+    MethodMatcher test();
+
+    static MethodMatcher levelTestMatcher(String typeDefinition, String level) {
+      return MethodMatcher.create()
+        .typeDefinition(typeDefinition)
+        .name(String.format("is%c%sEnabled", level.charAt(0), level.toLowerCase(Locale.ROOT).substring(2)))
+        .withoutParameter();
+    }
+
+    static Stream<LogLevels> logLevels() {
+      return Stream.concat(Arrays.stream(SLF4J_LEVELS.values()), Arrays.stream(JUL_LEVELS.values()));
+    }
+
+    enum SLF4J_LEVELS implements LogLevels {
+      TRACE,
+      DEBUG,
+      INFO,
+      WARN,
+      ERROR
+      ;
+
+      public static final TypeCriteria MARKER = TypeCriteria.is("org.slf4j.Marker");
+
+      @Override
+      public List<MethodMatcher> log() {
+        return slf4jVariants(() -> MethodMatcher.create().typeDefinition(SLF4J_LOGGER).name(toString().toLowerCase(Locale.ROOT)));
+      }
+
+      @Override
+      public MethodMatcher test() {
+        return levelTestMatcher(SLF4J_LOGGER, toString());
+      }
+
+      private static List<MethodMatcher> slf4jVariants(Supplier<MethodMatcher> prototype) {
+        return ImmutableList.of(
+          prototype.get().parameters(STRING),
+          prototype.get().parameters(STRING, TypeCriteria.anyType()),
```

Comment:
I would make a static import of `TypeCriteria.anyType()` to reduce a bit the verbosity of that block.